File: bite-route/server/src/domain.test.ts
import { describe, expect, it } from 'vitest';
import { canTransition, estimatePickupMinutes, pointsEarned, priceOrder } from './domain.js';

describe('Bite Route order rules', () => {
  it('prices promo, loyalty, tax and tip without moving money', () => {
    expect(priceOrder({subtotalCents:2500,promoPercent:10,loyaltyPointsRedeemed:100,tipCents:300,taxRate:.07})).toEqual({subtotalCents:2500,discountCents:250,loyaltyDiscountCents:100,taxCents:151,tipCents:300,totalCents:2601});
  });
  it('earns one point per dollar',()=>expect(pointsEarned(1499)).toBe(14));
  it('enforces order state transitions',()=>{expect(canTransition('received','preparing')).toBe(true);expect(canTransition('received','ready')).toBe(false);expect(canTransition('ready','completed')).toBe(true);});
  it('expands wait time as capacity fills',()=>{expect(estimatePickupMinutes(0,8)).toBe(15);expect(estimatePickupMinutes(9,8)).toBe(45);expect(estimatePickupMinutes(99,8,25)).toBe(25);});
});

File: bite-route/test/domain_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:bite_route/domain.dart';

void main() {
  test('modifier pricing and cart totals stay in integer cents', () {
    const line = CartLine(itemId: 'taco', name: 'Tacos', basePriceCents: 1350, quantity: 2, modifiers: [ModifierChoice('extra', 'Extra Beef', 300)]);
    expect(line.unitPriceCents, 1650);
    expect(line.lineTotalCents, 3300);
    final total = priceCart(lines: const [line], promoPercent: 10, loyaltyPointsRedeemed: 100, tipCents: 300, taxRate: .07);
    expect(total.subtotalCents, 3300);
    expect(total.discountCents, 330);
    expect(total.loyaltyDiscountCents, 100);
    expect(total.taxCents, 201);
    expect(total.totalCents, 3371);
  });

  test('loyalty earns one point per paid dollar', () => expect(pointsEarnedForOrder(1499), 14));

  test('order transitions protect kitchen workflow', () {
    expect(canTransitionOrder(OrderStatus.received, OrderStatus.preparing), isTrue);
    expect(canTransitionOrder(OrderStatus.received, OrderStatus.ready), isFalse);
    expect(canTransitionOrder(OrderStatus.ready, OrderStatus.completed), isTrue);
  });

  test('pickup slot grows with kitchen capacity', () {
    final now = DateTime(2026, 9, 13, 18, 1);
    expect(nextPickupSlot(now: now, openOrders: 0, capacityPer15Minutes: 8), DateTime(2026, 9, 13, 18, 30));
    expect(nextPickupSlot(now: now, openOrders: 9, capacityPer15Minutes: 8), DateTime(2026, 9, 13, 19, 0));
  });
}
